File: modules/apps/commerce/headless/headless-commerce/headless-commerce-delivery-cart-client/src/main/java/com/liferay/headless/commerce/delivery/cart/client/dto/v1_0/CartItem.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.headless.commerce.delivery.cart.client.dto.v1_0;

import com.liferay.headless.commerce.delivery.cart.client.function.UnsafeSupplier;
import com.liferay.headless.commerce.delivery.cart.client.serdes.v1_0.CartItemSerDes;

import java.io.Serializable;

import java.util.Map;
import java.util.Objects;

import javax.annotation.Generated;

/**
 * @author Andrea Sbarra
 * @generated
 */
@Generated("")
public class CartItem implements Cloneable, Serializable {

	public static CartItem toDTO(String json) {
		return CartItemSerDes.toDTO(json);
	}

	public CartItem[] getCartItems() {
		return cartItems;
	}

	public void setCartItems(CartItem[] cartItems) {
		this.cartItems = cartItems;
	}

	public void setCartItems(
		UnsafeSupplier<CartItem[], Exception> cartItemsUnsafeSupplier) {

		try {
			cartItems = cartItemsUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected CartItem[] cartItems;

	public Map<String, ?> getCustomFields() {
		return customFields;
	}

	public void setCustomFields(Map<String, ?> customFields) {
		this.customFields = customFields;
	}

	public void setCustomFields(
		UnsafeSupplier<Map<String, ?>, Exception> customFieldsUnsafeSupplier) {

		try {
			customFields = customFieldsUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Map<String, ?> customFields;

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public void setId(UnsafeSupplier<Long, Exception> idUnsafeSupplier) {
		try {
			id = idUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long id;

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public void setName(UnsafeSupplier<String, Exception> nameUnsafeSupplier) {
		try {
			name = nameUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String name;

	public String getOptions() {
		return options;
	}

	public void setOptions(String options) {
		this.options = options;
	}

	public void setOptions(
		UnsafeSupplier<String, Exception> optionsUnsafeSupplier) {

		try {
			options = optionsUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String options;

	public Long getParentCartItemId() {
		return parentCartItemId;
	}

	public void setParentCartItemId(Long parentCartItemId) {
		this.parentCartItemId = parentCartItemId;
	}

	public void setParentCartItemId(
		UnsafeSupplier<Long, Exception> parentCartItemIdUnsafeSupplier) {

		try {
			parentCartItemId = parentCartItemIdUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long parentCartItemId;

	public Price getPrice() {
		return price;
	}

	public void setPrice(Price price) {
		this.price = price;
	}

	public void setPrice(UnsafeSupplier<Price, Exception> priceUnsafeSupplier) {
		try {
			price = priceUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Price price;

	public Long getProductId() {
		return productId;
	}

	public void setProductId(Long productId) {
		this.productId = productId;
	}

	public void setProductId(
		UnsafeSupplier<Long, Exception> productIdUnsafeSupplier) {

		try {
			productId = productIdUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long productId;

	public Integer getQuantity() {
		return quantity;
	}

	public void setQuantity(Integer quantity) {
		this.quantity = quantity;
	}

	public void setQuantity(
		UnsafeSupplier<Integer, Exception> quantityUnsafeSupplier) {

		try {
			quantity = quantityUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Integer quantity;

	public Settings getSettings() {
		return settings;
	}

	public void setSettings(Settings settings) {
		this.settings = settings;
	}

	public void setSettings(
		UnsafeSupplier<Settings, Exception> settingsUnsafeSupplier) {

		try {
			settings = settingsUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Settings settings;

	public String getSku() {
		return sku;
	}

	public void setSku(String sku) {
		this.sku = sku;
	}

	public void setSku(UnsafeSupplier<String, Exception> skuUnsafeSupplier) {
		try {
			sku = skuUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String sku;

	public Long getSkuId() {
		return skuId;
	}

	public void setSkuId(Long skuId) {
		this.skuId = skuId;
	}

	public void setSkuId(UnsafeSupplier<Long, Exception> skuIdUnsafeSupplier) {
		try {
			skuId = skuIdUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Long skuId;

	public Boolean getSubscription() {
		return subscription;
	}

	public void setSubscription(Boolean subscription) {
		this.subscription = subscription;
	}

	public void setSubscription(
		UnsafeSupplier<Boolean, Exception> subscriptionUnsafeSupplier) {

		try {
			subscription = subscriptionUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected Boolean subscription;

	public String getThumbnail() {
		return thumbnail;
	}

	public void setThumbnail(String thumbnail) {
		this.thumbnail = thumbnail;
	}

	public void setThumbnail(
		UnsafeSupplier<String, Exception> thumbnailUnsafeSupplier) {

		try {
			thumbnail = thumbnailUnsafeSupplier.get();
		}
		catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	protected String thumbnail;

	@Override
	public CartItem clone() throws CloneNotSupportedException {
		return (CartItem)super.clone();
	}

	@Override
	public boolean equals(Object object) {
		if (this == object) {
			return true;
		}

		if (!(object instanceof CartItem)) {
			return false;
		}

		CartItem cartItem = (CartItem)object;

		return Objects.equals(toString(), cartItem.toString());
	}

	@Override
	public int hashCode() {
		String string = toString();

		return string.hashCode();
	}

	public String toString() {
		return CartItemSerDes.toJSON(this);
	}

}

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/mini_cart/CartItem.js
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import ClayButton from '@clayui/button';
import ClayIcon from '@clayui/icon';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, {useCallback, useContext, useState} from 'react';

import {PRODUCT_REMOVED_FROM_CART} from '../../utilities/eventsDefinitions';
import QuantitySelector from '../quantity_selector/QuantitySelector';
import ItemInfoView from './CartItemViews/ItemInfoView';
import ItemPriceView from './CartItemViews/ItemPriceView';
import MiniCartContext from './MiniCartContext';
import {parseOptions} from './util/index';

const REMOVAL_TIMEOUT = 2000,
	REMOVAL_CANCELING_TIMEOUT = 700,
	REMOVAL_ERRORS_TIMEOUT = 4000,
	INITIAL_ITEM_STATE = {
		isGettingRemoved: false,
		isRemovalCanceled: false,
		isRemoved: false,
		isShowingErrors: false,
		previousQuantity: null,
		removalTimeoutRef: null,
	};

function CartItem({item: cartItem}) {
	const {
		cartItems: childItems,
		errorMessages,
		id: cartItemId,
		name,
		options: rawOptions,
		price,
		quantity,
		settings,
		sku,
		skuId,
		thumbnail,
	} = cartItem;

	const {
		CartResource,
		cartState,
		displayDiscountLevels,
		setIsUpdating,
		spritemap,
		updateCartModel,
	} = useContext(MiniCartContext);

	const {id: orderId} = cartState;
	const [itemState, setItemState] = useState(INITIAL_ITEM_STATE);
	const [itemQuantity, setItemQuantity] = useState(quantity);
	const [itemPrice, updateItemPrice] = useState(price);

	const options = parseOptions(rawOptions);

	// eslint-disable-next-line react-hooks/exhaustive-deps
	const showErrors = () => {
		setIsUpdating(false);

		setItemState({
			...INITIAL_ITEM_STATE,
			isShowingErrors: true,
			removalTimeoutRef: setTimeout(() => {
				setItemState(INITIAL_ITEM_STATE);
			}, REMOVAL_ERRORS_TIMEOUT),
		});

		return Promise.resolve();
	};

	const cancelRemoveItem = () => {
		clearTimeout(itemState.removalTimeoutRef);

		setItemState({
			...INITIAL_ITEM_STATE,
			isRemovalCanceled: true,
			removalTimeoutRef: setTimeout(() => {
				setIsUpdating(false);

				setItemState(INITIAL_ITEM_STATE);
			}, REMOVAL_CANCELING_TIMEOUT),
		});
	};

	const removeItem = () => {
		setItemState({
			...INITIAL_ITEM_STATE,
			isGettingRemoved: true,
			removalTimeoutRef: setTimeout(() => {
				setIsUpdating(true);

				setItemState({
					...INITIAL_ITEM_STATE,
					isGettingRemoved: true,
					isRemoved: true,
					removalTimeoutRef: setTimeout(() => {
						CartResource.deleteItemById(cartItemId)
							.then(() => updateCartModel({orderId}))
							.then(() => {
								setIsUpdating(false);
								Liferay.fire(PRODUCT_REMOVED_FROM_CART, {
									skuId,
								});
							})
							.catch(showErrors);
					}, REMOVAL_CANCELING_TIMEOUT),
				});
			}, REMOVAL_TIMEOUT),
		});
	};

	const updateItemQuantity = useCallback(
		(quantity) => {
			if (quantity !== itemQuantity) {
				setIsUpdating(true);

				CartResource.updateItemById(cartItemId, {
					...cartItem,
					quantity,
				})
					.then(({quantity: updatedQuantity, ...updatedItem}) => {
						setItemQuantity(updatedQuantity);

						return Promise.resolve(updatedItem);
					})
					.then(({price: updatedPrice}) => {
						const {price: currentPriceValue} = itemPrice;
						const {price: updatedPriceValue} = updatedPrice;

						/**
						 * The unit price of an item may change based
						 * on the change of its quantity
						 * @type {boolean}
						 */

						const priceValueChanged =
							!currentPriceValue ||
							currentPriceValue !== updatedPriceValue;

						if (priceValueChanged) {
							return updateItemPrice(updatedPrice);
						}

						return Promise.resolve();
					})
					.then(() => updateCartModel({orderId}))
					.then(() => setIsUpdating(false))
					.catch(showErrors);
			}

			return Promise.resolve();
		}, // eslint-disable-next-line react-hooks/exhaustive-deps
		[CartResource, cartItem, cartItemId, orderId]
	);

	const {
		isGettingRemoved,
		isRemovalCanceled,
		isRemoved,
		isShowingErrors,
	} = itemState;

	return (
		<div
			className={classnames('mini-cart-item', isRemoved && 'is-removed')}
		>
			{!!thumbnail && (
				<div
					className={'mini-cart-item-thumbnail'}
					style={{backgroundImage: `url(${thumbnail})`}}
				/>
			)}

			<div
				className={classnames(
					'mini-cart-item-info',
					!!options && 'has-options'
				)}
			>
				<ItemInfoView
					childItems={childItems}
					name={name}
					options={options}
					sku={sku}
				/>
			</div>

			<div className={'mini-cart-item-quantity'}>
				<QuantitySelector
					onUpdate={updateItemQuantity}
					quantity={quantity}
					spritemap={spritemap}
					{...settings}
				/>
			</div>

			<div className={'mini-cart-item-price'}>
				<ItemPriceView
					displayDiscountLevels={displayDiscountLevels}
					price={itemPrice}
				/>
			</div>

			<div className={'mini-cart-item-delete'}>
				<button
					className={'btn btn-unstyled'}
					onClick={removeItem}
					type={'button'}
				>
					<ClayIcon
						spritemap={spritemap}
						symbol={'times-circle-full'}
					/>
				</button>
			</div>

			{errorMessages && (
				<div className={'mini-cart-item-errors'}>
					<ClayIcon
						spritemap={spritemap}
						symbol={'exclamation-circle'}
					/>

					<span>{errorMessages}</span>
				</div>
			)}

			{isShowingErrors && (
				<div className={'mini-cart-item-errors'}>
					<ClayIcon
						spritemap={spritemap}
						symbol={'exclamation-circle'}
					/>

					<span>
						{Liferay.Language.get('an-unexpected-error-occurred')}
					</span>
				</div>
			)}

			<div
				className={classnames(
					'mini-cart-item-removing',
					isGettingRemoved && 'active',
					isRemovalCanceled && 'canceled'
				)}
			>
				<span>{Liferay.Language.get('the-item-has-been-removed')}</span>
				<span>
					<ClayButton
						displayType={'link'}
						href={'#'}
						onClick={cancelRemoveItem}
						small
						type={'button'}
					>
						{Liferay.Language.get('undo')}
					</ClayButton>
				</span>
			</div>
		</div>
	);
}

CartItem.propTypes = {
	item: PropTypes.object,
};

export default CartItem;
